presto-main/src/main/java/com/facebook/presto/cost/HistogramCalculator.java

```diff
@@ -0,0 +1,167 @@
+        if (lowPercentile.isUnknown() || highPercentile.isUnknown()) {
+            if (!useHeuristics) {
+                return Estimate.unknown();
+            }
+            // in the case the histogram has no values
+            if (totalDistinctValues.equals(Estimate.zero()) || range.getDistinctValuesCount() == 0.0) {
+                return Estimate.of(0.0);
+            }
+
+            // in the case only one is unknown
+            if (((lowPercentile.isUnknown() && !highPercentile.isUnknown()) ||
+                    (!lowPercentile.isUnknown() && highPercentile.isUnknown())) &&
+                    isFinite(range.length())) {
+                return useHeuristics ? Estimate.of(StatisticRange.INFINITE_TO_FINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR) : Estimate.unknown();
+            }
+
+            if (range.length() == 0.0) {
+                return totalDistinctValues.map(distinct -> 1.0 / distinct);
+            }
+
+            if (!isNaN(range.getDistinctValuesCount())) {
+                return totalDistinctValues.map(distinct -> min(1.0, range.getDistinctValuesCount() / distinct));
+            }
+
+            return Estimate.of(StatisticRange.INFINITE_TO_INFINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR);
+        }
+
+        // in the case the range is a single value, this can occur if the input
+        // filter range is a single value (low == high) OR in the case that the
+        // bounds of the filter or this histogram are infinite.
+        // in the case of infinite bounds, we should return an estimate that
+        // correlates to the overlapping distinct values.
+        if (lowPercentile.equals(highPercentile)) {
+            if (!useHeuristics) {
+                return Estimate.zero();
+            }
+            // If one of the bounds is unknown, but both percentiles are equal,
+            // it's likely that a heuristic value was returned
+            if (max.isUnknown() || min.isUnknown()) {
+                return totalDistinctValues.flatMap(distinct -> lowPercentile.map(lowPercent -> distinct * lowPercent));
+            }
+
+            return totalDistinctValues.map(distinct -> 1.0 / distinct);
+        }
+
+        // in the case that we return the entire range, the returned factor percent should be
+        // proportional to the number of distinct values in the range
+        if (lowPercentile.equals(Estimate.zero()) && highPercentile.equals(Estimate.of(1.0)) && min.isUnknown() && max.isUnknown()) {
+            if (!useHeuristics) {
+                return Estimate.unknown();
+            }
+
+            if (totalDistinctValues.equals(Estimate.zero())) {
+                return Estimate.of(1.0);
+            }
+            return totalDistinctValues.flatMap(totalDistinct -> {
+                if (DoubleMath.fuzzyEquals(totalDistinct, 0.0, 1E-6)) {
+                    return Estimate.unknown();
+                }
```

Comment:
I'm a little confused here. I may miss some context message, but why return a `Estimate.of(1.0)` when `totalDistinctValues` equals 0, then return a `Estimate.unknown()` when `totalDistinctValues` fuzzy equals 0?

Reply:
This is a good point. I think originally I was defending against having a divisor of 0. I will update this to

```java
return totalDistinctValues.flatMap(totalDistinct -> {
                if (DoubleMath.fuzzyEquals(totalDistinct, 0.0, 1E-6)) {
                    return Estimate.of(1.0);
                }
                return Estimate.of(min(1.0, range.getDistinctValuesCount() / totalDistinct));
            })
            // in the case totalDistinct is NaN or 0
            .or(() -> Estimate.of(StatisticRange.INFINITE_TO_INFINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR));
```